src/integrated/soil_snow_model.jl

```diff
@@ -346,41 +346,132 @@ conditions, taking into account the presence of snow on the surface.
-    # influx = maximum possible rate of infiltration given precip, phase_change_flux, evaporation/condensation
-    # but if this exceeds infiltration capacity of the soil, runoff will
-    # be generated.
-    # Use top_bc.water as temporary storage to avoid allocation
-    influx = p.soil.top_bc.water
-    @. influx =
-        p.snow.water_runoff * p.snow.snow_cover_fraction +
+    # Liquid influx is a combination of precipitation and snowmelt in general
+    liquid_influx =
+        Soil.compute_liquid_influx(p, soil, prognostic_land_components)
+    # This partitions the influx into runoff and infiltration
+    Soil.update_infiltration_water_flux!(
+        p,
+        bc.runoff,
+        liquid_influx,
+        Y,
+        t,
+        soil,
+    )
+    # This computes the energy of the infiltrating water
+    infiltration_energy_flux = Soil.compute_infiltration_energy_flux(
+        p,
+        bc.runoff,
+        bc.atmos,
+        prognostic_land_components,
+        liquid_influx,
+        soil,
+        Y,
+        t,
+    )
+    # The actual boundary condition is a mix of liquid water infiltration and
+    # evaporation. The infiltration already has accounted for snow cover fraction,
+    # because the influx it is computed from has accounted for that.
+    # The last term, `excess water flux`, arises when snow melts in a timestep but
+    # has a nonzero sublimation which was applied for the entire step.
+    @. p.soil.top_bc.water =
+        p.soil.infiltration +
         p.excess_water_flux +
         (1 - p.snow.snow_cover_fraction) *
-        (p.soil.turbulent_fluxes.vapor_flux_liq + p.drivers.P_liq)
-    # The update_runoff! function computes how much actually infiltrates
-    # given influx and our runoff model bc.runoff, and updates
-    # p.soil.infiltration in place
-    Soil.Runoff.update_runoff!(p, bc.runoff, influx, Y, t, soil)
-    @. p.soil.top_bc.water = p.soil.infiltration
-
+        p.soil.turbulent_fluxes.vapor_flux_liq
     @. p.soil.top_bc.heat =
         (1 - p.snow.snow_cover_fraction) * (
             p.soil.R_n +
             p.soil.turbulent_fluxes.lhf +
             p.soil.turbulent_fluxes.shf
         ) +
         p.excess_heat_flux +
-        p.snow.snow_cover_fraction * p.ground_heat_flux
+        p.snow.snow_cover_fraction * p.ground_heat_flux +
+        infiltration_energy_flux
+
     return nothing
 end
 
+"""
+   compute_liquid_influx(p,
+                         model,
+                         prognostic_land_components::Val{(:snow, :soil,)},
+    ) 
+
+Returns the liquid water volume flux at the surface of the soil; this
+will then be partitioned into surface runoff and infiltration:
+influx = infiltration - runoff (runoff > 0; infiltration and influx < 0,
+and infiltration <= influx).
+
+In a model with snow as a prognostic variable, the influx is a combination
+of the liquid precipitation as a volume flux and the snowmelt leaving the 
+snowpack.
+"""

```

Comment:
Here and elsewhere: I think it would be better not to duplicate the docstrings. A structure that collects the different methods and then uses one docstring for all of them would be cleaner, I think.